improver/expected_value.py

```diff
@@ -0,0 +1,76 @@
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Calculation of expected value from a probability distribution."""
+
+import iris.analysis
+from iris.coords import CellMethod
+from iris.cube import Cube
+
+from improver import PostProcessingPlugin
+from improver.ensemble_copula_coupling.ensemble_copula_coupling import (
+    ConvertProbabilitiesToPercentiles,
+    RebadgePercentilesAsRealizations,
+)
+from improver.metadata.probabilistic import is_percentile, is_probability
+from improver.utilities.cube_manipulation import collapsed
+
+
+class ExpectedValue(PostProcessingPlugin):
+    """Calculation of expected value from a probability distribution"""
+
+    def process(self, cube: Cube) -> Cube:
+        """Expected value calculation and metadata updates.
+
+        Args:
+            cube:
+                Probabilistic data with a realization, threshold or percentile
+                representation.
+
+        Returns:
+            Expected value of probability distribution. Same shape as input cube
+            but with realization/threshold/percentile coordinate removed.
+        """
+        if is_probability(cube):
+            # TODO: replace this with direct calculation of the integral over
+            # probability thresholds. The current approach works and has the
+            # correct interface, but some accuracy will be lost in the conversion
+            # and memory usage is very high.
+            #
+            # 19 percentiles corresponds to 5, 10, 15...95%.
+            cube = ConvertProbabilitiesToPercentiles().process(
+                cube, no_of_percentiles=19
+            )
+        if is_percentile(cube):
+            cube = RebadgePercentilesAsRealizations().process(cube)
+        mean_cube = collapsed(cube, "realization", iris.analysis.MEAN)
+        mean_cube.remove_coord("realization")
```

Comment:
Just stumbled across this function: https://github.com/metoppv/improver/blob/d3598a6994f803a0af03d878b4e5d5d6a6ee73b1/improver/utilities/cube_manipulation.py#L84
which could be used here rather than `collapsed`.

Reply:
Duplicate of above.

Reply:
This seems to happen to me on occasion (duplicate items), I'm not sure why.